x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/scheduling/TransformScheduler.java

```diff
@@ -0,0 +1,236 @@
+        try {
+            taskWasProcessed = processScheduledTasksInternal();
+        } finally {
+            // Make sure we clear the "isProcessingActive" bit regardless of whether there was a success or failure.
+            // Otherwise, the processing would be stuck forever.
+            isProcessingActive.set(false);
+        }
+        Instant processingFinished = clock.instant();
+        logger.trace(
+            String.format(
+                Locale.ROOT,
+                "Processing scheduled tasks finished, took {}ms",
+                Duration.between(processingStarted, processingFinished).toMillis()
+            )
+        );
+        if (taskWasProcessed) {
+            // If we happened to process the task, there may be other tasks also eligible for processing.
+            // We try to process them ASAP as we don't want to wait the `delay` for every task.
+            processScheduledTasks();
+        }
+    }
+
+    private boolean processScheduledTasksInternal() {
+        if (scheduledTasks.isEmpty()) {
+            // There are no scheduled tasks, hence, nothing to do
+            return false;
+        }
+        long currentTimeMillis = clock.millis();
+        TransformScheduledTask scheduledTask = scheduledTasks.first();
+        // Check if the task is eligible for processing
+        if (currentTimeMillis < scheduledTask.getNextScheduledTimeMillis()) {
+            // It is too early to process this task.
+            // Consequently, it is too early to process other tasks because the tasks are sorted by their next scheduled time.
+            // Try again later.
+            return false;
+        }
+        // Create event that will be sent to the listener
+        Event event = new Event(scheduledTask.getTransformId(), scheduledTask.getNextScheduledTimeMillis(), currentTimeMillis);
+        // Trigger the listener
+        scheduledTask.getListener().triggered(event);
+        // Update the task's last_triggered_time to current time (next_scheduled_time gets automatically re-calculated)
+        scheduledTasks.update(scheduledTask.getTransformId(), task -> {
+            if (task.equals(scheduledTask) == false) {
+                logger.debug(
+                    () -> String.format(
+                        Locale.ROOT,
+                        "[{}] task object got modified while processing. Expected: {}, was: {}",
+                        scheduledTask.getTransformId(),
+                        scheduledTask,
+                        task
+                    )
+                );
+            }
+            return new TransformScheduledTask(
+                task.getTransformId(),
+                task.getFrequency(),
+                currentTimeMillis,
+                task.getFailureCount(),
+                task.getListener()
+            );
+        });
+        return true;
+    }
+
+    /**
+     * Stops the processing.
+     */
+    public void stop() {
+        threadPool.shutdown();
```

Comment:
Add a member `Scheduler.Cancellable scheduledFuture;` assign it in start() 

`scheduledFuture = threadPool.scheduleWithFixedDelay(..`

and cancel it here `scheduledFuture.cancel()`

Reply:
Sorry I was suggesting that as a replacement to `threadPool.shutdown();` Do you mean to shutdown the threadpool? Is it used again?

Reply:
Done.

Most likely shutting down the pool wasn't doing any harm as this was only called once during the `Transform` plugin `close()` method.
But since the scheduler gets the pool via an argument and does not own its lifetime, I agree it's cleaner to just cancel the task and **not** perform the shutdown in the `stop()` method.